components/application-mgt/org.wso2.carbon.identity.application.mgt/src/main/java/org/wso2/carbon/identity/application/mgt/listener/DefaultRoleManagementListener.java

```diff
@@ -0,0 +1,527 @@
+                        "Invalid audience. No application found with application id: " + applicationId);
+            }
+            boolean valid = false;
+            for (ServiceProviderProperty property : app.getSpProperties()) {
+                // TODO :  use osgi service to get this
+                if (ALLOWED_ROLE_AUDIENCE_PROPERTY_NAME.equals(property.getName()) &&
+                        APPLICATION.equalsIgnoreCase(property.getValue())) {
+                    valid = true;
+                }
+            }
+            if (!valid) {
+                throw new IdentityRoleManagementClientException(INVALID_AUDIENCE.getCode(),
+                        "Application: " + applicationId + " does not have Application role audience type");
+            }
+        } catch (IdentityApplicationManagementException e) {
+            String errorMessage = "Error while retrieving the application for the given id: " + applicationId;
+            throw new IdentityRoleManagementServerException(UNEXPECTED_SERVER_ERROR.getCode(), errorMessage, e);
+        }
+    }
+
+    /**
+     * Validate permissions for application audience.
+     *
+     * @param permissions Permissions.
+     * @param applicationId Application ID.
+     * @throws IdentityRoleManagementException Error occurred while validating permissions.
+     */
+    private void validatePermissionsForApplication(List<Permission> permissions, String applicationId,
+                                                   String tenantDomain)
+            throws IdentityRoleManagementException {
+
+        List<String> authorizedScopes = getAuthorizedScopes(applicationId, tenantDomain);
+        for (Permission permission : permissions) {
+            if (!authorizedScopes.contains(permission.getName())) {
+                throw new IdentityRoleManagementClientException(INVALID_PERMISSION.getCode(),
+                        "Permission : " + permission.getName() + " is not authorized for application " +
+                                "with ID: " + applicationId);
+            }
+        }
+    }
+
+    /**
+     * Get authorized scopes by app ID.
+     *
+     * @param appId App ID.
+     * @param tenantDomain Tenant domain.
+     * @throws IdentityRoleManagementException Error occurred while retrieving authorized scopes by app ID.
+     */
+    private List<String> getAuthorizedScopes(String appId, String tenantDomain)
+            throws IdentityRoleManagementException {
+
+        AuthorizedAPIManagementService authorizedAPIManagementService = new AuthorizedAPIManagementServiceImpl();
+        List<AuthorizedScopes> authorizedScopesList;
+        try {
+            authorizedScopesList = authorizedAPIManagementService.getAuthorizedScopes(appId, tenantDomain);
+        } catch (IdentityApplicationManagementException e) {
+            throw new IdentityRoleManagementException("Error while retrieving authorized scopes.",
```

Comment:
shouldn't is chaged to (errorcode, message) format?

Reply:
fixed